tests/integration/test_s3.py

```diff
@@ -2692,3 +2595,112 @@ def test_replay_s3_call(api_version, bucket_name, payload):
 
     bucket_head = s3_client.head_bucket(Bucket=bucket_name)
     assert bucket_head["ResponseMetadata"]["HTTPStatusCode"] == 200
+
+
+@patch.object(config, "DISABLE_CUSTOM_CORS_S3", False)
+def test_cors_with_single_origin_error(s3_client):
+    # works with TEST_TARGET=AWS_CLOUD
+    bucket_cors_config = {
+        "CORSRules": [
+            {
+                "AllowedOrigins": ["https://localhost:4200"],
+                "AllowedMethods": ["GET", "PUT"],
+                "MaxAgeSeconds": 3000,
+                "AllowedHeaders": ["*"],
+            }
+        ]
+    }
+
+    bucket_name = "bucket-%s" % short_uid()
+    object_key = "424f6bae-c48f-42d8-9e25-52046aecc64d/document.pdf"
+    s3_client.create_bucket(Bucket=bucket_name)
+    s3_client.put_bucket_cors(Bucket=bucket_name, CORSConfiguration=bucket_cors_config)
+
+    # create signed url
+    url = s3_client.generate_presigned_url(
+        ClientMethod="put_object",
+        Params={
+            "Bucket": bucket_name,
+            "Key": object_key,
+            "ContentType": "application/pdf",
+            "ACL": "bucket-owner-full-control",
+        },
+        ExpiresIn=3600,
+    )
+    result = requests.put(
+        url,
+        data="something",
+        verify=False,
+        headers={
+            "Origin": "https://localhost:4200",
+            "Content-Type": "application/pdf",
+        },
+    )
+    assert result.status_code == 200
+
+    bucket_cors_config = {
+        "CORSRules": [
+            {
+                "AllowedOrigins": [
+                    "https://localhost:4200",
+                    "https://localhost:4201",
+                ],
+                "AllowedMethods": ["GET", "PUT"],
+                "MaxAgeSeconds": 3000,
+                "AllowedHeaders": ["*"],
+            }
+        ]
+    }
+
+    s3_client.put_bucket_cors(Bucket=bucket_name, CORSConfiguration=bucket_cors_config)
+
+    # create signed url
+    url = s3_client.generate_presigned_url(
+        ClientMethod="put_object",
+        Params={
+            "Bucket": bucket_name,
+            "Key": object_key,
+            "ContentType": "application/pdf",
+            "ACL": "bucket-owner-full-control",
+        },
+        ExpiresIn=3600,
+    )
+
+    # mimic chrome behavior, sending OPTIONS request first for strict-origin-when-cross-origin
```

Comment:
this is mostly original `TestS3.test_cors_with_single_origin_error`. The most notable addition this `OPTIONS` request.